File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    // Memoization cache: dp[day][transactionsLeft][holdingStock]
    // dp[i][j][k] represents max profit at day i with j transactions left and k=1 if holding stock
    private Integer[][][] dp;
    private int[] prices;
    private int n;

    public int maxProfit(int k, int[] prices) {
        n = prices.length;
        this.prices = prices;
        // Initialize memoization array
        // n days, k+1 transactions (0 to k), 2 states (0: not holding, 1: holding)
        dp = new Integer[n][k + 1][2];
      
        // Start from day 0, with k transactions available, not holding any stock
        return dfs(0, k, 0);
    }

    /**
     * Dynamic programming with memoization to find maximum profit
     * @param day Current day index
     * @param transactionsLeft Number of complete transactions remaining
     * @param isHolding 0 if not holding stock, 1 if holding stock
     * @return Maximum profit from current state
     */
    private int dfs(int day, int transactionsLeft, int isHolding) {
        // Base case: no more days left
        if (day >= n) {
            return 0;
        }
      
        // Return memoized result if already computed
        if (dp[day][transactionsLeft][isHolding] != null) {
            return dp[day][transactionsLeft][isHolding];
        }
      
        // Option 1: Do nothing on this day (skip to next day)
        int maxProfit = dfs(day + 1, transactionsLeft, isHolding);
      
        if (isHolding == 1) {
            // Currently holding stock: can sell
            // Selling completes a transaction, so transactionsLeft stays the same
            maxProfit = Math.max(maxProfit, 
                                prices[day] + dfs(day + 1, transactionsLeft, 0));
        } else if (transactionsLeft > 0) {
            // Not holding stock and have transactions left: can buy
            // Buying starts a new transaction, so decrement transactionsLeft
            maxProfit = Math.max(maxProfit, 
                                -prices[day] + dfs(day + 1, transactionsLeft - 1, 1));
        }
      
        // Memoize and return the result
        return dp[day][transactionsLeft][isHolding] = maxProfit;
    }
}

File: leetcode/0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.ts
/**
 * Calculates the maximum profit from at most k stock transactions
 * @param k - Maximum number of transactions allowed
 * @param prices - Array of stock prices on each day
 * @returns Maximum profit achievable
 */
function maxProfit(k: number, prices: number[]): number {
    const n: number = prices.length;
  
    // Initialize memoization table
    // memo[day][transactionsRemaining][holdingStock]
    // memo[i][j][0] = max profit on day i with j transactions left, not holding stock
    // memo[i][j][1] = max profit on day i with j transactions left, holding stock
    const memo: number[][][] = Array.from({ length: n }, () =>
        Array.from({ length: k + 1 }, () => 
            Array.from({ length: 2 }, () => -1)
        )
    );
  
    /**
     * Dynamic programming helper function with memoization
     * @param day - Current day index
     * @param transactionsRemaining - Number of transactions still available
     * @param isHoldingStock - 1 if currently holding stock, 0 otherwise
     * @returns Maximum profit from current state onwards
     */
    const dfs = (day: number, transactionsRemaining: number, isHoldingStock: number): number => {
        // Base case: no more days to trade
        if (day >= n) {
            return 0;
        }
      
        // Return memoized result if already computed
        if (memo[day][transactionsRemaining][isHoldingStock] !== -1) {
            return memo[day][transactionsRemaining][isHoldingStock];
        }
      
        // Option 1: Do nothing on this day
        let maxProfitFromHere: number = dfs(day + 1, transactionsRemaining, isHoldingStock);
      
        if (isHoldingStock) {
            // Option 2: Sell stock today (complete a transaction)
            maxProfitFromHere = Math.max(
                maxProfitFromHere, 
                prices[day] + dfs(day + 1, transactionsRemaining, 0)
            );
        } else if (transactionsRemaining > 0) {
            // Option 2: Buy stock today (start a new transaction)
            maxProfitFromHere = Math.max(
                maxProfitFromHere, 
                -prices[day] + dfs(day + 1, transactionsRemaining - 1, 1)
            );
        }
      
        // Memoize and return the result
        memo[day][transactionsRemaining][isHoldingStock] = maxProfitFromHere;
        return maxProfitFromHere;
    };
  
    // Start from day 0, with k transactions available, not holding any stock
    return dfs(0, k, 0);
}
